### PR TITLE
FSE: New Patterns

### PR DESCRIPTION
**Pull request by ianstewart**
This PR should add 23 new patterns to the FSE Plugin.

This is _most_ of the patterns from …

https://dotcompatterns.wordpress.com/2020/05/13/quote/

through to …

https://dotcompatterns.wordpress.com/2020/06/16/audio-player/

with a few exceptions where I encountered a bug or small known issue. (We'll get to them in a later PR or via another deployment method.)

### CODE REVIEW

File: apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/images-02.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+		esc_html__( 'Summer', 'full-site-editing' ),
+		esc_html__( '2019', 'full-site-editing' ),
+		esc_html__( 'Failed Encounter', 'full-site-editing' ),
+		esc_html__( '2019', 'full-site-editing' ),
+		esc_html__( 'Movement', 'full-site-editing' ),
+		esc_html__( '2018', 'full-site-editing' ),
+		esc_html__( 'Retratos', 'full-site-editing' ),
+		esc_html__( '2017', 'full-site-editing' ),
+		esc_html__( 'Disappearance', 'full-site-editing' ),
+		esc_html__( '2016', 'full-site-editing' ),
```
Comment (andrewserong) [inline apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/images-02.php:97]: This is throwing a PHPCS error (trailing comma in a function call)

File: apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/images.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+		esc_html__( 'The fearsomeness of the wood had been a good bit rubbed off for me by Master Case’s banjo-strings and graven images, yet I thought it was a dreary walk, and guessed, when the disciples went up there, they must be badly scared. The light of the lantern, striking among all these trunks and forked branches and twisted rope-ends of lianas, made the whole place, or all that you could see of it, a kind of a puzzle of turning shadows. They came to meet you, solid and quick like giants, and then span off and vanished; they hove up over your head like clubs, and flew away into the night like birds. The floor of the bush glimmered with dead wood, the way the match-box used to shine after you had struck a lucifer. Big, cold drops fell on me from the branches overhead like sweat. There was no wind to mention; only a little icy breath of a land-breeze that stirred nothing; and the harps were silent.', 'full-site-editing' )
```
Comment (andrewserong) [inline apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/images.php:69]: This is throwing a couple of PHPCS whitespace errors:

```
 69 | ERROR | [x] Spaces must be used for mid-line alignment; tabs
    |       |     are not allowed
    |       |     (WordPress.WhiteSpace.DisallowInlineTabs.NonIndentTabsUsed)
 69 | ERROR | [x] Whitespace found at end of line
    |       |     (Squiz.WhiteSpace.SuperfluousWhitespace.EndLine)
```

File: apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/three-quotes.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	'categories' => array( 'quotes', 'text', 'list' ),
+	'content'    => sprintf(
+		$markup,
+		esc_html__( 'What Our Customers are Saying', 'full-site-editing' ),
+		esc_html__( 'Beverly Mattingdale', 'full-site-editing' ),
+		esc_html__( '“The audio quality is noticeably more amazing than everything else in my studio!”', 'full-site-editing' ),
+		esc_html__( 'Jefferson Thorpe', 'full-site-editing' ),
+		esc_html__( '“I was instantly blown away by their sound design and consistency.”', 'full-site-editing' ),
+		esc_html__( 'Calvin Tristan', 'full-site-editing' ),
+		esc_html__( '“Utterly impressive—nothing more needs to be said about these folks.”', 'full-site-editing')
```
Comment (andrewserong) [inline apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/three-quotes.php:90]: This is throwing a PHPCS whitespace error (space before closing bracket)

**Comment by iamtakashi** [comment]
> Some of the designs are quite tall, which doesn't feel like the best format for displaying them

The narrow default viewport size for the preview is making patterns with text a lot taller. Let's add `'viewportWidth' => 1280` to each pattern like https://github.com/Automattic/wp-calypso/pull/43362

File: apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/images-02.php (hunk 1 above)
Comment (ianstewart) [reply apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/images-02.php:97]: Derp. I'll fix it. :)

File: apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/images.php (hunk 2 above)
Comment (ianstewart) [reply apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/images.php:69]: Derp. Will fix.

File: apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/three-quotes.php (hunk 3 above)
Comment (ianstewart) [reply apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/three-quotes.php:90]: Derp. Got it.

**Comment by ianstewart** [comment]
The last commit should address the viewport and the non-standards code. Doesn't yet address any Atomic issues.

**Comment by ianstewart** [comment]
> Are we reaching a point where we need a new interface for browsing, filtering or selecting the patterns? 

There are some improvements on the way in core Gutenberg for surfacing patterns on inline search in the canvas and some designs out there for at least two column browsing.

**Comment by ianstewart** [comment]
> a fair few of the blocks are failing rendering in my Atomic site. E.g. List, Three Quotes, Heading and Text in Two Columns, Multi-column Text with Headline, Image and Text Mosaic, Subscription.

@iamtakashi any commonalities jumping out for you with this list?

**Comment by iamtakashi** [comment]
In my testing, the following patterns are broken in Simple sites too.

```
heading-and-text-in-two-columns.php
image-and-text-mosaic.php
three-quotes.php
multi-column-text-with-headline.php
list-02.php
subscription.php
```

In addition to the above, `subscription-02.php` is broken in Atomic sites. Though, this pattern appears fine in Simple sites. 

> any commonalities jumping out for you with this list?

It might be something to do with the column block.  The following patterns share the use of the column block

```
heading-and-text-in-two-columns.php
image-and-text-mosaic.php
three-quotes.php
multi-column-text-with-headline.php
```

I'm not sure why `list-02.php` and `subscription.php` are broken.

**Comment by deBhal** [comment]
Ok, the first problem is that the `%"` in "flex-basis:x%" is being interpreted as an sprintf formatting command. That's a bit of a head-scratcher, because `%"` is _not_ a printf formatting command - https://www.php.net/manual/en/function.printf.php

My best guess is that it's invalid syntax, terminating the printf flag but consuming the character in the process:
`printf( '>%"< >%#< >%y<. The d: >%d<' . "\n", '1', '2','3', '4');` =>
`>< >< ><. The d: >4<`

Interestingly, printf does complain if you don't provide an argument for those invalid flags, but still probably not a very useful behavioural quirk. I had no idea about any of this - :TIL:. 

Short version: `sed -i '' 's/%"/%%"/g' *.php`

If we `grep -l '%"' *`, we get:
```
heading-and-text-in-two-columns.php
image-and-text-mosaic.php
multi-column-text-with-headline.php
subscription.php
three-quotes.php
```

Which only misses `list-02.php`. I'll check that out next.

**Comment by deBhal** [comment]
The list pattern doesn't quite match the result of the save() function - there's one opening and one closing `<p>` tag that differ between what we've got in the repo and what gutenberg is generating on my AT site.

I've just pushed the change, as it seems the easiest way to communicate clearly here, but it needs testing in more scenarios. Let me know if you need to me to circle back to do that tomorrow.

Out of interest, where did these `$markup`s come from? Are we looking at different Gutenberg versions here, or just something like a copy-and-paste error? I understood that the patterns generally live as posts on a blog, so if we're pulling them out of a blog like that, we should have a look at that tool.

There seems to be another intermittent issue where the list-02 does not receive the list items, which results in an error in the console and an empty example in the inserter, but the actual block when you click on it is just fine.

It's a bit late here, so I'll come back to that tomorrow unless someone knows the story there.

I'm also still seeing a bunch of console errors when opening up the inserter after applying the percent fix I mentioned above, so I won't push that until I've had a chance to look at them tomorrow.

**Comment by ianstewart** [comment]
> Out of interest, where did these $markups come from? Are we looking at different Gutenberg versions here, or just something like a copy-and-paste error? I understood that the patterns generally live as posts on a blog, so if we're pulling them out of a blog like that, we should have a look at that tool.

Correct. It's dotcompatterns.wordpress.com.

File: apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/list-02.php [hunk 4]
```diff
@@ -14,7 +14,7 @@
 <!-- /wp:spacer -->
 
 <!-- wp:paragraph {"style":{"typography":{"fontSize":72,"lineHeight":"0.9"}}} -->
-<p style="line-height:.9;font-size:72px;"><p style="line-height:.9;font-size:72px;"><strong>%1$s<br>%2$s<br>%3$s </strong><br><strong>%4$s<br>%5$s<br>%6$s<br>%7$s<br>%8$s<br>%9$s</strong></p></p>
+<p style="line-height:.9;font-size:72px;"></p><p style="line-height:.9;font-size:72px;"><strong>%1$s<br>%2$s<br>%3$s </strong><br><strong>%4$s<br>%5$s<br>%6$s<br>%7$s<br>%8$s<br>%9$s</strong></p><p></p>
```
Comment (iamtakashi) [inline apps/full-site-editing/full-site-editing-plugin/block-patterns/patterns/list-02.php:17]: Wouldn't this make two empty paragraphs and make it three paragraphs in total?

The original markup was also weird though, it looks like a paragraph wrapped in a paragraph. And looking at the inline style, the paragraph tag `<p style="line-height:.9;font-size:72px;">` has been duplicated in the markup for some reason?

**Comment by iamtakashi** [comment]
The particular pattern was made in here: https://wordpress.com/block-editor/post/dotcompatterns.wordpress.com/938

I saw a duplicated paragraph tag in the markup. Not sure how it got there, but I cleaned it up. So if we update the markup in `list-02.php` with the current markup that has only a single paragraph tag, the pattern should be fine.

**Comment by iamtakashi** [comment]
I spoke too soon. The duplicate tag comes back when I reopen the pattern.

This: 
`<p style="line-height:0.9;font-size:72px" ... </p>` 

becomes this, and it seems like a bug.
`<p style="line-height:0.9;font-size:72px"><p style="line-height:.9;font-size:72px;"> ... </p></p>`

**Comment by deBhal** [comment]
The notifications in the console and editor are triggered when the HTML regenerated by Gutenberg doesn’t match what’s on the page, so if there’s a bug in one of the component blocks, Gutenberg will complain if we fix it.

What should be able to unit-test this, I’ll have a look at writing that. It will probably be quicker than the manual testing I was planning, and should catch this whole class of problems in future

**Comment by ianstewart** [comment]
cc @ramonjd just in case any of the above turns out to be useful for the Patterns API work.

**Comment by ockham** [comment]
There are still a number of phpcs lints, all of which seem to be fixable through phpcbf. I wonder why our pre-commit hook isn't doing that for us? :thinking: https://github.com/Automattic/wp-calypso/blob/8e806f2ce055e30db6a3eb7e627d052ce48c7169/bin/pre-commit-hook.js#L135-L154

**Comment by ockham** [comment]
I'll rebase and fix those lints.

**Comment by ockham** [comment]
BTW what's the state of this PR? Does the block markup need more checking/fixing, or is it ready to go?

**Comment by ockham** [comment]
> I spoke too soon. The duplicate tag comes back when I reopen the pattern.
> 
> This:
> `<p style="line-height:0.9;font-size:72px" ... </p>`
> 
> becomes this, and it seems like a bug.
> `<p style="line-height:0.9;font-size:72px"><p style="line-height:.9;font-size:72px;"> ... </p></p>`

This seems vaguely familar. Is this taken directly from the dotcompatterns site, or run through any additional tooling? Specifically something that involves an XML or HTML parser (block patterns API something something)?

I seem to recall a problem with those parsers that would change multi-node markup (`<abcd>foo</abcd><efgh>bar</efgh>`) into single-node markup -- IIRC by wrapping everything in a duplicate of the first node (`<abcd><abcd>foo</abcd><efgh>bar</efgh></abcd>`). (The solution was to make sure that the parser only got single-node markup in the first place -- if needed, by wrapping everything in a root node that could be removed after parsing.)

cc/ @deBhal

**Comment by ianstewart** [comment]
> BTW what's the state of this PR? Does the block markup need more checking/fixing, or is it ready to go?

I missed this message on Monday, sorry @ockham. I'll need some help testing on Atomic sites but I'll test again on Simple and check the status there. If the patterns load w/o errors we can merge and ship!

**Comment by ianstewart** [comment]
Hmm. Four of these are patterns are not rendering correctly on my Simple test site.

image

**Comment by lsl** [comment]
> `<p><p></p></p>` turning into `<p></p><p></p>`

That looks a lot like what I was seeing with the DOMDocument parser, it will emit errors but those are usually being tossed in most cases and the parser just repairs the html the best it can.

WordPress/gutenberg/pull/24645 or D48290-code may have fixed those for this case, if there is actually invalid html in the pattern source that probably needs to just be fixed.

**Comment by david-szabo97** [comment]
- Moved files from the old FSE folder to the new ET folder.
- Fixed an argument typo in **Podcast Subscription**
- Fixed markup in **List**
- Escaped percents in **Heading and Text in Two Columns**, **Image and Text Mosaic**, **Multi-column Text with Headline**, ** Subscription**, and **Three Quotes**

They are all looking good on local. Can someone help me out by testing the patterns?

**Comment by roo2** [comment]
hmm this all looks good except for the youtube embed block, on my simple site, the youtube embed blocks comes up with this error:
Screen Shot 2020-09-03 at 10 38 37 AM
After attempting block recovery, the block works as expected

**Comment by roo2** [comment]
I fixed that youtube bug and will deploy this once tests pass
